src/sqlfluff/core/rules/base.py

```diff
@@ -484,149 +567,107 @@ def _eval(self, context: RuleContext) -> EvalResultType:
-                        "https://github.com/sqlfluff/sqlfluff/issues ?\n"
-                        "You can ignore this exception for now, by adding "
-                        f"'-- noqa: {self.code}' at the end\n"
-                        f"of line {exception_line}\n"
-                    ),
+        vs: List[SQLLintError] = []
+        fixes: List[LintFix] = []
+
+        # Propagates memory from one rule _eval() to the next.
+        memory: Any = root_context.memory
+        context = None
+        crawl_behavior = CrawlBehavior(
+            self._works_on_unparsable, self.recurse_into, self.needs_raw_stack
+        )
+        for context in crawl_behavior.crawl(root_context):
+            try:
+                context.memory = memory
+                res = self._eval(context=context)
+            except (bdb.BdbQuit, KeyboardInterrupt):  # pragma: no cover
+                raise
+            # Any exception at this point would halt the linter and
+            # cause the user to get no results
+            except Exception as e:
+                self.logger.critical(
+                    f"Applying rule {self.code} threw an Exception: {e}", exc_info=True
                 )
-            )
-            return vs, raw_stack, fixes, memory
-
-        new_lerrs: List[SQLLintError] = []
-        new_fixes: List[LintFix] = []
-
-        if res is None:
-            # Assume this means no problems (also means no memory)
-            pass
-        elif isinstance(res, LintResult):
-            # Extract any memory
-            memory = res.memory
-            self._adjust_anchors_for_fixes(context, res)
-            self._process_lint_result(
-                res, templated_file, ignore_mask, new_lerrs, new_fixes
-            )
-        elif isinstance(res, list) and all(
-            isinstance(elem, LintResult) for elem in res
-        ):
-            # Extract any memory from the *last* one, assuming
-            # it was the last to be added
-            memory = res[-1].memory
-            for elem in res:
-                self._adjust_anchors_for_fixes(context, elem)
+                assert context.segment.pos_marker
+                exception_line, _ = context.segment.pos_marker.source_position()
+                self._log_critical_errors(e)
+                vs.append(
+                    SQLLintError(
+                        rule=self,
+                        segment=context.segment,
+                        fixes=[],
+                        description=(
+                            f"Unexpected exception: {str(e)};\n"
+                            "Could you open an issue at "
+                            "https://github.com/sqlfluff/sqlfluff/issues ?\n"
+                            "You can ignore this exception for now, by adding "
+                            f"'-- noqa: {self.code}' at the end\n"
+                            f"of line {exception_line}\n"
+                        ),
+                    )
```

Comment:
Might be nice to move this sort of _safe_exec logic into a wrapper of some sort.
such that one could focus on the loop logic.